os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -303,58 +292,145 @@ static int binary_manager_load_all(void)
 	return BINMGR_OPERATION_FAIL;
 }
 
-static void reload_kill_each(FAR struct tcb_s *tcb, FAR void *arg)
+#ifdef CONFIG_BINMGR_RECOVERY
+/****************************************************************************
+ * Name: binary_manager_reload
+ *
+ * Description:
+ *   This function will terminate all the task/thread created by the binary
+ *   i.e input binary id.
+ *   It is called after all children are excluded from scheduling by fault recovery.
+ *   It terminates its children, unloads binary and then it will load the binary.
+ *
+ * Input parameters:
+ *   binid   -   The pid of binary to be reload
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+static int binary_manager_reload(int binid)
 {
 	int ret;
-	int binid;
+	int bin_idx;
+	struct tcb_s *tcb;
+	struct tcb_s *btcb;
 
-	binid = (int)arg;
-	if (binid < 0) {
-		return;
+	/* Check whether it is registered in binary manager */
+	bin_idx = binary_manager_get_index_with_binid(binid);
+	if (bin_idx < 0) {
+		bmdbg("binary %s is not registered\n", binid);
+		return BINMGR_NOT_FOUND;
 	}
 
-	if (tcb->group->tg_loadtask == binid && tcb->pid != binid) {
-		ret = task_terminate(tcb->pid, true);
-		if (ret < 0) {
-			bmdbg("Failed to terminate %d\n", tcb->pid);
+	/* Terminate all children created in a binary in a inactive list */
+	tcb = (struct tcb_s *)dq_peek((dq_queue_t *)g_tasklisttable[TSTATE_TASK_INACTIVE].list);
```

Comment:
We can change this to do~while. Then we can reduce the duplicated code.

Reply:
I think it can't be changed because next tcb is got from dq_peek or dq_next according to tcb value.

Reply:
Okay. But this logic looks inefficient. Sometimes it tries to access multiple times for some nodes.
I think we can just use each binaries linked list.